rfcs/0003-multi-blockchain-support.md

```diff
@@ -0,0 +1,534 @@
+      ['address', 'Address', code => `${code}.toAddress()`],
+      ['bool', 'boolean', code => `${code}.toBoolean()`],
+      ['byte', 'Bytes', code => `${code}.toBytes()`],
+      [
+        /^bytes(1|2|3|4|5|6|7|8|9|10|11|12|13|14|15|16|17|18|19|20|21|22|23|24|25|26|27|28|29|30|31|32)?$/,
+        'Bytes',
+        code => `${code}.toBytes()`,
+      ],
+      ...
+    ]
+  },
+  
+  // Code generation for ABIs,
+  codegen: (abi) => {
+    // Return a string to be put into the output file for the ABI
+  }
+}
+```
+
+Mapping files ending with `.ts` are automatically detected wherever they appear
+in the manifest and included in builds and deployments.
+
+#### Graph TS
+
+Each chain supported in mappings comes with its own `chains/<name>.ts` file in
+the `graph-ts` repository. Each of these has at least two sections: one for
+declaring host exports that are required for mappings to interact with the
+chain, another to define chain-specific types and helpers written in
+AssemblyScript. The file `chains/ethereum.ts` file could look as follows:
+
+```js
+import { Address } from '../index'
+
+export declare namespace ethereum {
+  function call(contract: Address, fn: string, args: Array<Value>): void
+}
+
+export namespace ethereum {
+    class Value { ... }
+    class Block { ... }
+    class Transaction { ... }
+    class Event { ... }
+    class Contract { ... }
+}
+```
+
+The main `index.ts` in the `graph-ts` repository re-exports each of these chain
+integrations via
+
+```js
+export * from './chains/ethereum'
+export * from './chains/substrate'
+```
+
+So that they can be imported into subgraph mappings with
+
+```js
+import { ethereum, substrate } from './chains/substrate'
+```
+
+## Compatibility
+
+The proposed design is mostly backwards-compatible, except for a few areas:
+
+- The manifest changes require a subgraph manifest migration in Graph CLI, as
+  well as database migration for the subgraph meta data to match the new
+  structure.
+
+- Moving AssemblyScript types from e.g. `EthereumValue` to `ethereum.Value`
+  _may_ break subgraphs that directly access these types. This is ok, because
+  the majority (if not all) subgraphs only use them indirectly through generated
+  code. This does not affect subgraphs that are already deployed.
+
+## Drawbacks and Risks
+
+Generally, supporting more blockchains is an opportunity much more than it is a
+risk. It widens the reach of The Graph to communities beyond Ethereum.
+
+One risk in the design itself is that it may require a bit of experimentation to
+get the Graph Node traits for integrating arbitrary chains right. Overall, the
+work resulting from this RFC may easily cover one or two months; note that
+estimating this work is the responsibility of the engineering plan, however.
+
+## Alternatives
+
+There are nuances where the design could look different, specifically around how
+the manifest spec could be updated and whether new chain integrations are developed
+as plugins or not.
```

Comment:
I would like to suggest we avoid doing plugin development for the initial MVP. The plugin design is likely to overcomplicate this change because it brings into play issues like backward compatibility and generally broadens the surface area of the design. Integrating a couple of chains in-house will give an opportunity for the finer details of the design to solidify, giving a clearer picture of the best interface for plugins.

It's much easier to add plugin support later and move our in-house integrations to plugins then it will be to make any necessary changes to an ad-hoc plugin system in the future. This is especially true given the lack of a clear direction on the best way to tackle plugins in the Rust community.

Reply:
Yep, I'm 100% with you there.